tests/e2e/configurer/configurer.go

```diff
@@ -0,0 +1,513 @@
+	"github.com/osmosis-labs/osmosis/v7/tests/e2e/util"
+)
+
+type Configurer interface {
+	ConfigureChains() error
+
+	ClearResources() error
+
+	GetChainConfig(chainIndex int) ChainConfig
+
+	RunSetup() error
+
+	RunValidators() error
+
+	RunIBC() error
+
+	SendIBC(srcChain *chain.Chain, dstChain *chain.Chain, recipient string, token sdk.Coin)
+
+	CreatePool(chainId string, valIdx int, poolFile string)
+}
+
+type BaseConfigurer struct {
+	chainConfigs   []*ChainConfig
+	dockerImages   *dockerImages.ImageConfig
+	dockerPool     *dockertest.Pool
+	dockerNetwork  *dockertest.Network
+	valResources   map[string][]*dockertest.Resource
+	hermesResource *dockertest.Resource
+	setupTests     setupFn
+	t              *testing.T
+}
+
+type status struct {
+	LatestHeight string `json:"latest_block_height"`
+}
+
+type syncInfo struct {
+	SyncInfo status `json:"SyncInfo"`
+}
+
+const (
+	// osmosis version being upgraded to (folder must exist here https://github.com/osmosis-labs/osmosis/tree/main/app/upgrades)
+	UpgradeVersion = "v9"
+	// estimated number of blocks it takes to submit for a proposal
+	PropSubmitBlocks float32 = 10
+	// estimated number of blocks it takes to deposit for a proposal
+	PropDepositBlocks float32 = 10
+	// number of blocks it takes to vote for a single validator to vote for a proposal
+	PropVoteBlocks float32 = 1.2
+	// number of blocks used as a calculation buffer
+	PropBufferBlocks float32 = 5
+	// max retries for json unmarshalling
+	MaxRetries = 60
+)
+
+var (
+	// whatever number of validator configs get posted here are how many validators that will spawn on chain A and B respectively
+	validatorConfigsChainA = []*chain.ValidatorConfig{
+		{
+			Pruning:            "default",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+		{
+			Pruning:            "nothing",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+		{
+			Pruning:            "custom",
+			PruningKeepRecent:  "10000",
+			PruningInterval:    "13",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+		{
+			Pruning:            "everything",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   0,
+			SnapshotKeepRecent: 0,
+		},
+	}
+	validatorConfigsChainB = []*chain.ValidatorConfig{
+		{
+			Pruning:            "default",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+		{
+			Pruning:            "nothing",
+			PruningKeepRecent:  "0",
+			PruningInterval:    "0",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+		{
+			Pruning:            "custom",
+			PruningKeepRecent:  "10000",
+			PruningInterval:    "13",
+			SnapshotInterval:   1500,
+			SnapshotKeepRecent: 2,
+		},
+	}
+)
+
+func New(t *testing.T, isIBCEnabled, isUpgradeEnabled bool) (Configurer, error) {
```

Comment:
Here we are using a factory design pattern to decide what kind of Configurer to use.

- if only ibc is enabled, we want to have 2 chains initialized at the local version of Osmosis codebase
- If only upgrade is enabled, that's invalid (we can decouple upgrade testing from IBC in a future PR)
- if both are enabled, we want 2 chains with IBC initialized at the previous Osmosis version
- if non are enabled, we only need one chain at the local version of Osmosis code

Similarly, we can add logic for things like state sync here in the future

Reply:
Can we convert this github comment to a code comment on the function?